### DISCUSSION TITLE
MQTT blocking connect mqttStart() behavior

### DISCUSSION THREAD

**Discussion by sleepypig2003** [HASwitchPlate/openHASP#187] [General]
I noticed that the behavior when openhasp lost the mqtt connection is that it clears the UI, stops everything and displays a progress bar.

I think when this is running on a switchplate with a local relay, like the lanbon8, this would prevent users from operating the local relay (turn on/off lights), when homeassistant is off.

Is there a way we can have mqttStart() to immediately exit if the connection fails?  This way openhasp just keeps retrying.

The two other changes I am thinking of:
1. Allow a deconfiguration of the mqtt server, so openhasp stops trying to query homeassistant
2. And perhaps having the reboot timeout optional as well in the future, since I can think of some situations where homeassistant can be offline indefinitely 

hasp_mqtt_pubsubclient.cpp:
void mqttStart()
...
    haspProgressMsg(F(D_MQTT_CONNECTING));
    haspProgressVal(mqttReconnectCount * 5);
...

what do you think if I put a return on failure?  ie.

...
            default:
                LOG_WARNING(TAG_MQTT, F("Unknown failure"));
                return;
        }

ps. I also noticed there are threaded mqtt code in openhasp - unsure if this was the intended direction this was going to go to

**Comment by fvanroie** [comment]
There is an ongoing discussion about this here:   Allow offline mode #61

For arduino we use Pubsubclient. On the PC we use Paho single threaded as mqtt client.
The Multi threaded asynchronous paho client is not being used.

Pubsubclient has a blocking timeout when it's trying to (re)connect to the MQTT broker.
This is annoying when this happens at the time you're using the plate.
Ideally the Pubsubclient should be swapped for an async MQTT client and/or multi-threaded implementation...

There's just not an overall solution for this issue yet. A great deal of the functionality depends on the broker being active.
